includes/class-wc-payments-order-service.php

```diff
@@ -0,0 +1,732 @@
+	 *
+	 * @return string Transaction details page url.
+	 */
+	private function compose_transaction_url( $charge_id ) {
+		if ( empty( $charge_id ) ) {
+			return '';
+		}
+
+		return add_query_arg(
+			[
+				'page' => 'wc-admin',
+				'path' => '/payments/transactions/details',
+				'id'   => $charge_id,
+			],
+			admin_url( 'admin.php' )
+		);
+	}
+
+	/**
+	 * Composes url for dispute details page.
+	 *
+	 * @param string $dispute_id Dispute id.
+	 *
+	 * @return string Dispute details page url.
+	 */
+	private function compose_dispute_url( $dispute_id ) {
+		return add_query_arg(
+			[
+				'page' => 'wc-admin',
+				'path' => '/payments/disputes/details',
+				'id'   => $dispute_id,
+			],
+			admin_url( 'admin.php' )
+		);
+	}
+
+	/**
+	 * Check if order is locked for payment processing
+	 *
+	 * @param WC_Order $order  The order that is being paid.
+	 * @param string   $intent_id The id of the intent that is being processed.
+	 *
+	 * @return bool    A flag that indicates whether the order is already locked.
+	 */
+	private function is_order_locked( $order, $intent_id = null ) {
+		$order_id       = $order->get_id();
+		$transient_name = 'wcpay_processing_intent_' . $order_id;
+		$processing     = get_transient( $transient_name );
+
+		// Block the process if the same intent is already being handled.
+		return ( '-1' === $processing || ( isset( $intent_id ) && $processing === $intent_id ) );
+	}
+
+	/**
+	 * Lock an order for payment intent processing for 5 minutes.
+	 *
+	 * @param WC_Order $order     The order that is being paid.
+	 * @param string   $intent_id The id of the intent that is being processed.
+	 *
+	 * @return void
+	 */
+	private function lock_order_payment( $order, $intent_id = null ) {
+		$order_id       = $order->get_id();
+		$transient_name = 'wcpay_processing_intent_' . $order_id;
+
+		set_transient( $transient_name, empty( $intent_id ) ? '-1' : $intent_id, 5 * MINUTE_IN_SECONDS );
+	}
+
+	/**
+	 * Unlocks an order for processing by payment intents.
+	 *
+	 * @param WC_Order $order The order that is being unlocked.
+	 *
+	 * @return void
+	 */
+	private function unlock_order_payment( $order ) {
+		$order_id = $order->get_id();
+		delete_transient( 'wcpay_processing_intent_' . $order_id );
+	}
+
+	/**
+	 * Refreshes the order from the database, checks if it is locked, and locks it.
+	 *
+	 * TODO: Update to throw exceptions so try/catch can be used.
+	 * TODO: Maybe add checks to see if there is already a successful intent, or the intent status passed is already set.
+	 *
+	 * @param WC_Order $order   Order object.
+	 * @param string   $intent_id The ID of the intent associated with this order.
+	 *
+	 * @return bool
+	 */
+	private function order_prepared_for_processing( $order, $intent_id ) {
+		if ( ! is_a( $order, 'WC_Order' ) ) {
+			return false;
+		}
+
+		// Read the latest order properties from the database to avoid race conditions if webhook was handled during this request.
+		$clone_order = clone $order;
```

Comment:
This looks good, do you think it's worth extracting this part to a private function in case we need to reuse it anywhere else? I was thinking something like `is_order_paid()`.

Reply:
I think this is a good idea. Were you thinking like this?

```
	private function order_prepared_for_processing( $order, $intent_id ) {
		if ( ! is_a( $order, 'WC_Order' ) ) {
			return false;
		}

		if ( $this->is_order_paid( $order ) ) {
			return false;
		}

		if ( $this->is_order_locked( $order, $intent_id ) ) {
			return false;
		}

		// Lock the order.
		$this->lock_order_payment( $order, $intent_id );

		return true;
	}

	/**
	 * Checks to see if the current order, and a fresh copy of the order from the database are paid.
	 *
	 * @param WC_Order $order The order being checked.
	 *
	 * @return boolean True if it has a paid status, false if not.
	 */
	private function is_order_paid( $order ) {
		// Read the latest order properties from the database to avoid race conditions if webhook was handled during this request.
		$clone_order = clone $order;
		$clone_order->get_data_store()->read( $clone_order );

		// Check if the order is already complete.
		if ( function_exists( 'wc_get_is_paid_statuses' ) ) {
			if ( $order->has_status( wc_get_is_paid_statuses() )
				|| $clone_order->has_status( wc_get_is_paid_statuses() ) ) {
				return true;
			}
		}

		return false;
	}
```

Reply:
Yep, that's exactly what I had in mind!